proposals/010-authorizer.md

```diff
@@ -0,0 +1,515 @@
+     * The implementation must ensure that the returned authorization partitions all the given {@code actions}
+     * between {@link AuthorizeResult#allowed()} and {@link AuthorizeResult#denied()}.
+     * @param subject The subject.
+     * @param actions The actions.
+     * @return The outcome.
+     */
+    CompletionStage<AuthorizeResult> authorize(Subject subject, List<Action> actions);
+    
+    /**
+     * All the resource types supported by this authorizer.
+     */
+    Set<Class<?>> resourceTypes();
+}
+```
+
+Note:
+* The asynchronous return type is intended to permit implementations using technologies such as [OPA][OPA] or [OpenFGA][OpenFGA], which would otherwise need to block the thread making the decision over the network.
+* Because of the possibility of higher latency authorizer implementations, call sites should try to pass as many `actions` as possible so as to benefit from batching requests.
+* Unlike Apache Kafka we don't currently include in the Authorizer the responsibility to audit log authorization attempts.
+
+An `Action` describes something the subject may be attempting to do:
+
+```java
+public record Action(
+                     ResourceType<?> resourceType,
+                     String resourceName) { ... }
+```
+
+Where `ResourceType` has this declaration:
+
+```java
+/**
+ * Interface to be implemented by {@code enum} types which represent a type of resource, such as a 
+ * Kafka topic or consumer group. The enum elements represent the operations that instances of that resource type support.
+ * For example a Kafka topic supports a CREATE, DESCRIBE, READ and WRITE operations (among others).
+ */
+public interface ResourceType<V extends Enum<V> & ResourceType<V>> { ... }
+```
+
+The scary-looking type parameter bound means a `ResourceType` implementation must be an `enum` rather than `class`, `interface`, etc.
+The `enum`'s elements are the operations that a particular resource type supports. 
+We would model a Kafka topic like this:
+
+```java
+enum Topic implements ResourceType<Topic> {
+    READ,
+    WRITE,
+    CREATE,
+    // ...
+}
+
+The `java.lang.Class` of a particular `ResourceType` then serves to identify the type of a resource in the API.
+In this way, the `Action` instance for describing a topic called "foo" would be created like so:
+
+```java
+new Action(Topic.DESCRIBE, "foo")
+```
+
+The `AuthorizeResult` type provides access to the decisions about each of the actions in the query to `Authorizer.authorize()`.
+That is, `Authorizer.authorize()` partitions its `actions` argument into the `allowed` and `denied` lists of the returned `AuthorizeResult`.
+
+```java
+public enum Decision {
+    ALLOW,
+    DENY;
+}
+
+public class AuthorizeResult {
+                            
+	AuthorizeResult(
+		Subject subject,
+		List<Action> allowed,
+		List<Action> denied) {
+        // ...
+    }
+
+    public Decision decision(Operation<?> operation, String resourceName) { ... }
```

Comment:
I think this is the only mention of Operation. Should it be `public Decision decision(Action action)`?

Reply:
@robobario `ResourceType` was at one point known as `Operation`, and this one obviously escaped by limited powers of search and replace. Sorry about that.